src/Corrade/configure.h.cmake

```diff
@@ -403,4 +403,22 @@ static_assert(sizeof(1 ? "" : "") == 1,
 #endif
 #endif
 
+#if CORRADE_CXX_STANDARD >= 202002L
+#define CORRADE_CXX20_CONSTEXPR constexpr
+#else
+#define CORRADE_CXX20_CONSTEXPR
+#endif
+
+#if CORRADE_CXX_STANDARD >= 201703L
+#define CORRADE_CXX17_CONSTEXPR constexpr
+#else
+#define CORRADE_CXX17_CONSTEXPR
+#endif
+
+#if CORRADE_CXX_STANDARD >= 201402L
+#define CORRADE_CXX14_CONSTEXPR constexpr
+#else
+#define CORRADE_CXX14_CONSTEXPR
+#endif
```

Comment:
There's already [CORRADE_CONSTEXPR14](https://doc.magnum.graphics/corrade/Macros_8h.html#af57746d07f1b55515249dd8f7f1d9e11) in `Corrade/Utility/Macros.h`, and I'd like the other two to follow a similar naming and be in the same header.

But also please see my comment in mosra/magnum#597 about having a special macro for C++14+ constexpr only if a "is constant evaluated" builtin is available.

Reply:
Alright. Seems like my original proposal wasn't sufficient for what we're trying to achieve. It's not too useful for any of `CORRADE_CONSTEXPR{17,20,23}` to be defined -- C++14 relaxed constexpr is already enough for anything Vector has to handle. But `CORRADE_CONSTEXPR14` needs to reflect whether it works in practice (`__cpp_constexpr` is accurate here). Then we have several kinds of constexpr functions:

- C++11 semantics -- `constexpr`
- constexpr with C++14 rules, no need for inline asm or intrinsics -- `CORRADE_CONSTEXPR14`
- ones that need SIMD -- `CORRADE_CONSTEVAL14`, #ifndef `CORRADE_CONSTEVAL`

The last case can be used like that:

```c++
template<typename T>
CORRADE_CONSTEVAL14 inline T foo(T x)
{
#ifdef CORRADE_CONSTEVAL
    if CORRADE_CONSTEVAL
    {
        // do something traditional to 'x'
    }
    else
#endif
    {
        // do something funny to 'x'
    }
}
```

If you want to enable constexpr when lacking `CORRADE_CONSTEVAL` but when SIMD has been disabled, then it could be set to `constexpr(true)` as long as `__cpp_if_constexpr >= 201606` holds (aka C++17).

Reply:
I think it's better to remove `CORRADE_CONSTEVAL14` from this pull request and put it in Magnum with a name such as `MAGNUM_VECTOR_CONSTEVAL14`. It's too specialized for a global definition. Here it is for reference:

```c++
#ifdef CORRADE_CONSTEVAL
#define CORRADE_CONSTEVAL14 CORRADE_CONSTEXPR14
#else
#define CORRADE_CONSTEVAL14
#endif
```